playground/utils/kaolin_future/interpolated_cameras.py

```diff
@@ -0,0 +1,293 @@
+    height = round(height_1 * (1 - x) + height_2 * x)
+    intrinsics = {intr_name: intr_val[0] * (1 - x) + intr_val[1] * x for intr_name, intr_val in intrinsics.items()}
+
+    cam = Camera.from_args(
+        eye=eye,
+        at=eye - target,
+        up=up,
+        width=width, height=height,
+        device=cam1.device,
+        **intrinsics
+    )
+
+    return cam
+
+
+def interpolate_camera_on_spline_path(
+    trajectory: List[Camera],
+    timestep: int,
+    frames_between_cameras: int = 60
+) -> Camera:
+    r"""Interpolates a camera from a linear path formed by a trajectory of cameras.
+    The trajectory is assumed to have a list of at least 4 cameras.
+    The interpolation is done using Catmull-Rom Splines.
+
+    Args:
+        trajectory (List[kaolin.render.camera.Camera]): A trajectory of camera nodes, used to form a continuous path.
+        timestep (int): Timestep used to interpolate a point on the smoothed trajectory.
+            `timestep` can take any integer value to support continuous animations.
+            e.g. if `timestep > len(trajectory) X frames_between_cameras`, the timestep will fold over using a
+            modulus.
+        frames_between_cameras (int): Number of interpolated points generated between each pair of cameras on the
+            trajectory. In essence, this value controls how detailed, or smooth the path is.
+
+    Returns:
+        (kaolin.render.camera.Camera):
+            An interpolated camera object formed by the cameras trajectory.
+    """
+    traj_idx = (timestep // frames_between_cameras) % len(trajectory)
+
+    traj_idx = min(max(traj_idx, 0), len(trajectory) - 3)
+
+    cam1 = trajectory[traj_idx - 1]
+    cam2 = trajectory[traj_idx]
+    cam3 = trajectory[traj_idx + 1]
+    cam4 = trajectory[traj_idx + 2]
+
+    eye_1, target_1, up_1 = cam1.cam_pos(), cam1.cam_forward(), cam1.cam_up()
+    eye_2, target_2, up_2 = cam2.cam_pos(), cam2.cam_forward(), cam2.cam_up()
+    eye_3, target_3, up_3 = cam3.cam_pos(), cam3.cam_forward(), cam3.cam_up()
+    eye_4, target_4, up_4 = cam4.cam_pos(), cam4.cam_forward(), cam4.cam_up()
+    width_1, height_1 = cam1.width, cam1.height
+    width_2, height_2 = cam2.width, cam2.height
+    width_3, height_3 = cam3.width, cam3.height
+    width_4, height_4 = cam4.width, cam4.height
+
+    intrinsics = dict()
+    if cam1.lens_type == 'pinhole' and cam2.lens_type == 'pinhole':
+        intrinsics['fov'] = (
+            cam1.fov(in_degrees=False),
+            cam2.fov(in_degrees=False),
+            cam3.fov(in_degrees=False),
+            cam4.fov(in_degrees=False)
+        )
+    elif cam1.lens_type == 'ortho' and cam2.lens_type == 'ortho':
+        intrinsics['fov_distance'] = \
+            (cam1.fov_distance(), cam2.fov_distance(), cam3.fov_distance(), cam4.fov_distance())
+
+    Xs = np.linspace(0.0, 1.0, frames_between_cameras)
+    t = Xs[timestep % frames_between_cameras]
+
+    eye = _catmull_rom(t, eye_1, eye_2, eye_3, eye_4)
```

Comment:
using some kind of lie algebra / quaternion interpolation rather than vector interpolation probably makes sense here as well